File: Week 18/cpp/1.DynamicProgramming.cpp
/*
 * WEEK 18 - C++ DSA
 * Topic: Dynamic Programming Fundamentals
 * File: 1.DynamicProgramming.cpp
 *
 * CONCEPT:
 *     DP solves a problem by decomposing it into overlapping subproblems and
 *     storing each sub-answer to avoid recomputation. Two preconditions:
 *         1. Optimal substructure.
 *         2. Overlapping subproblems.
 *     Two implementation styles:
 *         TOP-DOWN  : recursion + memo (std::vector cache or std::unordered_map).
 *         BOTTOM-UP : tabulate from base case upward (usually faster, often
 *                     space-optimisable).
 *
 * KEY POINTS:
 *     - Identify state, recurrence, base case, and evaluation order.
 *     - Space-optimise when the recurrence depends only on a few prior rows.
 *     - For DP on subsets/knapsack-style, iterating capacity in REVERSE
 *       on the 1-D rolling array prevents reusing the current item.
 *
 * ALGORITHM / APPROACH:
 *     Five canonical problems:
 *         1. 0/1 Knapsack             (2-D and space-optimised 1-D)
 *         2. Longest Common Subseq.   (2-D)
 *         3. Coin change (min coins)  (1-D)
 *         4. Subset sum               (1-D boolean)
 *         5. Climbing stairs          (Fibonacci, O(1) state)
 *
 * C++-SPECIFIC NOTES:
 *     - std::vector<std::vector<int>> for 2-D DP tables.
 *     - INT_MAX from <climits>, or use a large sentinel; std::min works.
 *     - Pass arrays by const& to avoid copies.
 *     - For very large tables, consider a 1-D vector indexed manually
 *       (i * (W+1) + w) for cache-friendliness.
 *
 * DRY RUN:
 *     Knapsack weights={2,3,4,5}, values={3,4,5,6}, W=5
 *         dp[4][5] = 7  (take items with w=2,v=3 and w=3,v=4 -> total v=7)
 *
 *     LCS("ABCBDAB","BDCABA") -> 4
 *
 *     coinChange([1,2,5], 11)
 *         dp = [0,1,1,2,2,1,2,2,3,3,2,3]; answer = 3 (5+5+1).
 *
 * COMPLEXITY:
 *     knapsack          O(n*W) time, O(n*W) space (O(W) optimised)
 *     lcs               O(m*n) time, O(m*n) space
 *     coinChange        O(target * |coins|) time, O(target) space
 *     subsetSum         O(n * target) time, O(target) space
 *     climbStairs       O(n) time, O(1) space
 */

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <climits>

using std::vector;
using std::string;
using std::max;
using std::min;

// ---------- 0/1 KNAPSACK ----------
int knapsack(const vector<int>& weights, const vector<int>& values, int W) {
    int n = (int)weights.size();
    vector<vector<int>> dp(n + 1, vector<int>(W + 1, 0));
    for (int i = 1; i <= n; ++i) {
        int wi = weights[i - 1], vi = values[i - 1];
        for (int w = 0; w <= W; ++w) {
            dp[i][w] = dp[i - 1][w];
            if (wi <= w)
                dp[i][w] = max(dp[i][w], dp[i - 1][w - wi] + vi);
        }
    }
    return dp[n][W];
}

int knapsackOptimised(const vector<int>& weights, const vector<int>& values, int W) {
    vector<int> dp(W + 1, 0);
    for (size_t i = 0; i < weights.size(); ++i) {
        for (int w = W; w >= weights[i]; --w)               // REVERSE!
            dp[w] = max(dp[w], dp[w - weights[i]] + values[i]);
    }
    return dp[W];
}

// ---------- LONGEST COMMON SUBSEQUENCE ----------
int lcs(const string& s1, const string& s2) {
    int m = (int)s1.size(), n = (int)s2.size();
    vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));
    for (int i = 1; i <= m; ++i)
        for (int j = 1; j <= n; ++j)
            if (s1[i - 1] == s2[j - 1])
                dp[i][j] = dp[i - 1][j - 1] + 1;
            else
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1]);
    return dp[m][n];
}

// ---------- COIN CHANGE (min coins) ----------
int coinChange(const vector<int>& coins, int target) {
    const int INF = target + 1;                              // sentinel
    vector<int> dp(target + 1, INF);
    dp[0] = 0;
    for (int a = 1; a <= target; ++a)
        for (int c : coins)
            if (c <= a) dp[a] = min(dp[a], dp[a - c] + 1);
    return dp[target] > target ? -1 : dp[target];
}

// ---------- SUBSET SUM ----------
bool subsetSum(const vector<int>& nums, int target) {
    vector<bool> dp(target + 1, false);
    dp[0] = true;
    for (int x : nums)
        for (int t = target; t >= x; --t)
            dp[t] = dp[t] || dp[t - x];
    return dp[target];
}

// ---------- CLIMBING STAIRS ----------
int climbStairs(int n) {
    if (n <= 2) return n;
    int prev2 = 1, prev1 = 2;
    for (int i = 3; i <= n; ++i) {
        int curr = prev1 + prev2;
        prev2 = prev1;
        prev1 = curr;
    }
    return prev1;
}

int main() {
    std::cout << "=== 0/1 Knapsack ===\n";
    vector<int> weights{2,3,4,5}, values{3,4,5,6};
    std::cout << "W=5  bottom-up: " << knapsack(weights, values, 5) << "\n";   // 7
    std::cout << "W=10 bottom-up: " << knapsack(weights, values, 10) << "\n";  // 13
    std::cout << "W=10 optimised: " << knapsackOptimised(weights, values, 10) << "\n";

    std::cout << "\n=== LCS ===\n";
    std::cout << "LCS(ABCBDAB,BDCABA) = " << lcs("ABCBDAB", "BDCABA") << "\n"; // 4
    std::cout << "LCS(AGGTAB,GXTXAYB) = " << lcs("AGGTAB", "GXTXAYB") << "\n"; // 4

    std::cout << "\n=== Coin Change ===\n";
    std::cout << "coins=[1,2,5], amount=11: " << coinChange({1,2,5}, 11) << "\n"; // 3
    std::cout << "coins=[2],     amount=3 : " << coinChange({2}, 3) << "\n";       // -1
    std::cout << "coins=[1],     amount=0 : " << coinChange({1}, 0) << "\n";       // 0

    std::cout << "\n=== Subset Sum ===\n";
    std::cout << std::boolalpha;
    std::cout << "{3,1,5,9,12} target=8: " << subsetSum({3,1,5,9,12}, 8) << "\n"; // true
    std::cout << "{3,1,5,9,12} target=7: " << subsetSum({3,1,5,9,12}, 7) << "\n"; // false

    std::cout << "\n=== Climbing Stairs ===\n";
    for (int i = 1; i <= 10; ++i)
        std::cout << "climbStairs(" << i << ") = " << climbStairs(i) << "\n";
}

/*
 * NOTES (C++ vs Java):
 *   - Java auto-initialises int arrays to 0; std::vector<int>(n, 0) is the
 *     C++ analogue.
 *   - INT_MAX overflow is a real risk in C++ when adding 1 — we use
 *     (target + 1) as a saturating sentinel instead.
 *   - std::vector<bool> is bit-packed (special template) — fine for our
 *     boolean DP table; for raw speed prefer std::vector<char>.
 *   - Pass big inputs by const& to avoid quadratic copying.
 *   - For very tight memory, switch to a flat 1-D buffer (vector<int>) and
 *     compute indices manually.
 */

File: Week 18/rust/1.DynamicProgramming.rs
/*
 * WEEK 18 - RUST DSA
 * Topic: Dynamic Programming Fundamentals
 * File: 1.DynamicProgramming.rs
 *
 * CONCEPT:
 *     DP solves a problem by decomposing it into overlapping subproblems and
 *     storing each sub-answer. Two preconditions:
 *         1. Optimal substructure.
 *         2. Overlapping subproblems.
 *     Two implementation styles:
 *         TOP-DOWN  : recursion + memo (Vec or HashMap as cache).
 *         BOTTOM-UP : tabulation from base case upward (often faster, often
 *                     space-optimisable).
 *
 * KEY POINTS:
 *     - Identify state, recurrence, base case, evaluation order.
 *     - For 0/1 knapsack on a 1-D rolling array, iterate capacity in REVERSE
 *       so an item isn't reused.
 *     - Use i32::MAX or a "target+1" sentinel for "impossible" — Rust will
 *       panic on overflow in debug builds, so add carefully.
 *
 * ALGORITHM / APPROACH:
 *     Same five canonical problems as the Java reference:
 *         1. 0/1 Knapsack             (2-D and space-optimised 1-D)
 *         2. Longest Common Subseq.   (2-D)
 *         3. Coin change (min coins)  (1-D)
 *         4. Subset sum               (1-D boolean)
 *         5. Climbing stairs          (Fibonacci, O(1) state)
 *
 * RUST-SPECIFIC NOTES:
 *     - `vec![vec![0; W+1]; n+1]` is the canonical 2-D DP table.
 *     - For a top-down memo, the borrow checker rules out `RefCell<HashMap>`
 *       free-form recursion; idiomatic Rust prefers an explicit memo passed
 *       as `&mut Vec<...>` or simply iterative tabulation.
 *     - `usize` everywhere for indices; cast i32 with `as usize` carefully.
 *     - String slices use bytes() for ASCII inputs (LCS over ASCII strings).
 *
 * DRY RUN:
 *     Knapsack weights=[2,3,4,5], values=[3,4,5,6], W=5
 *         Final dp[4][5] = 7.
 *
 *     LCS("ABCBDAB","BDCABA") -> 4.
 *
 *     coin_change([1,2,5], 11) -> 3 (5+5+1).
 *
 *     subset_sum([3,1,5,9,12], 7) -> false; target=8 -> true.
 *
 * COMPLEXITY:
 *     knapsack          O(n*W) time, O(n*W) space (O(W) optimised)
 *     lcs               O(m*n) time, O(m*n) space
 *     coin_change       O(target * |coins|) time, O(target) space
 *     subset_sum        O(n * target) time, O(target) space
 *     climb_stairs      O(n) time, O(1) space
 */

use std::collections::HashMap;

// ---------- 0/1 KNAPSACK ----------
pub fn knapsack(weights: &[i32], values: &[i32], capacity: i32) -> i32 {
    let n = weights.len();
    let w_cap = capacity as usize;
    let mut dp = vec![vec![0i32; w_cap + 1]; n + 1];
    for i in 1..=n {
        let wi = weights[i - 1] as usize;
        let vi = values[i - 1];
        for w in 0..=w_cap {
            dp[i][w] = dp[i - 1][w];
            if wi <= w {
                let take = dp[i - 1][w - wi] + vi;
                if take > dp[i][w] { dp[i][w] = take; }
            }
        }
    }
    dp[n][w_cap]
}

pub fn knapsack_optimised(weights: &[i32], values: &[i32], capacity: i32) -> i32 {
    let w_cap = capacity as usize;
    let mut dp = vec![0i32; w_cap + 1];
    for i in 0..weights.len() {
        let wi = weights[i] as usize;
        let vi = values[i];
        for w in (wi..=w_cap).rev() {                         // REVERSE!
            let take = dp[w - wi] + vi;
            if take > dp[w] { dp[w] = take; }
        }
    }
    dp[w_cap]
}

// Top-down memo using HashMap (state = (i, capacity))
pub fn knapsack_memo(weights: &[i32], values: &[i32], capacity: i32) -> i32 {
    fn go(weights: &[i32], values: &[i32], i: usize, cap: i32,
          memo: &mut HashMap<(usize, i32), i32>) -> i32 {
        if i == weights.len() || cap == 0 { return 0; }
        if let Some(&v) = memo.get(&(i, cap)) { return v; }
        let skip = go(weights, values, i + 1, cap, memo);
        let take = if weights[i] <= cap {
            go(weights, values, i + 1, cap - weights[i], memo) + values[i]
        } else { 0 };
        let best = skip.max(take);
        memo.insert((i, cap), best);
        best
    }
    let mut memo = HashMap::new();
    go(weights, values, 0, capacity, &mut memo)
}

// ---------- LONGEST COMMON SUBSEQUENCE ----------
pub fn lcs(s1: &str, s2: &str) -> i32 {
    let a = s1.as_bytes();
    let b = s2.as_bytes();
    let m = a.len();
    let n = b.len();
    let mut dp = vec![vec![0i32; n + 1]; m + 1];
    for i in 1..=m {
        for j in 1..=n {
            dp[i][j] = if a[i - 1] == b[j - 1] {
                dp[i - 1][j - 1] + 1
            } else {
                dp[i - 1][j].max(dp[i][j - 1])
            };
        }
    }
    dp[m][n]
}

// ---------- COIN CHANGE (min coins) ----------
pub fn coin_change(coins: &[i32], target: i32) -> i32 {
    let t = target as usize;
    let inf = target + 1;                                     // sentinel
    let mut dp = vec![inf; t + 1];
    dp[0] = 0;
    for a in 1..=t {
        for &c in coins {
            let cu = c as usize;
            if c <= a as i32 && dp[a - cu] + 1 < dp[a] {
                dp[a] = dp[a - cu] + 1;
            }
        }
    }
    if dp[t] > target { -1 } else { dp[t] }
}

// ---------- SUBSET SUM ----------
pub fn subset_sum(nums: &[i32], target: i32) -> bool {
    let t = target as usize;
    let mut dp = vec![false; t + 1];
    dp[0] = true;
    for &x in nums {
        if x < 0 || x as usize > t { continue; }
        let xu = x as usize;
        for cap in (xu..=t).rev() {
            if dp[cap - xu] { dp[cap] = true; }
        }
    }
    dp[t]
}

// ---------- CLIMBING STAIRS ----------
pub fn climb_stairs(n: u32) -> u64 {
    if n <= 2 { return n as u64; }
    let (mut prev2, mut prev1) = (1u64, 2u64);
    for _ in 3..=n {
        let curr = prev1 + prev2;
        prev2 = prev1;
        prev1 = curr;
    }
    prev1
}

fn main() {
    println!("=== 0/1 Knapsack ===");
    let weights = [2, 3, 4, 5];
    let values  = [3, 4, 5, 6];
    println!("W=5  bottom-up: {}", knapsack(&weights, &values, 5));            // 7
    println!("W=10 bottom-up: {}", knapsack(&weights, &values, 10));           // 13
    println!("W=10 optimised: {}", knapsack_optimised(&weights, &values, 10));
    println!("W=10 memoised : {}", knapsack_memo(&weights, &values, 10));

    println!("\n=== LCS ===");
    println!("LCS('ABCBDAB','BDCABA') = {}", lcs("ABCBDAB", "BDCABA"));        // 4
    println!("LCS('AGGTAB','GXTXAYB') = {}", lcs("AGGTAB", "GXTXAYB"));        // 4

    println!("\n=== Coin Change ===");
    println!("coins=[1,2,5], amount=11: {}", coin_change(&[1, 2, 5], 11));     // 3
    println!("coins=[2],     amount=3 : {}", coin_change(&[2], 3));            // -1
    println!("coins=[1],     amount=0 : {}", coin_change(&[1], 0));            // 0

    println!("\n=== Subset Sum ===");
    println!("[3,1,5,9,12] target=8: {}", subset_sum(&[3, 1, 5, 9, 12], 8));   // true
    println!("[3,1,5,9,12] target=7: {}", subset_sum(&[3, 1, 5, 9, 12], 7));   // false

    println!("\n=== Climbing Stairs ===");
    for i in 1..=10u32 {
        println!("climb_stairs({:2}) = {}", i, climb_stairs(i));
    }
}

/*
 * NOTES (Rust vs Java):
 *   - `vec![vec![0; W+1]; n+1]` is the safe way to build a 2-D Vec; the
 *     macro clones the inner row each time (no aliasing pitfall).
 *   - i32 overflow panics in debug builds — picking a sentinel like
 *     `target + 1` is safer than i32::MAX when we later add 1.
 *   - Rust's top-down memo needs an explicit HashMap or &mut Vec; there's
 *     no @cache decorator. Iterative tabulation is usually preferred.
 *   - LCS over &str must convert to bytes (s.as_bytes()) for cheap indexing
 *     on ASCII; for full Unicode use s.chars().collect::<Vec<_>>().
 *   - climb_stairs uses u64 to avoid overflow for large n (Fibonacci grows fast).
 */
